core/src/main/scala/scalaz/Foldable.scala

```diff
@@ -325,6 +325,9 @@ trait Foldable[F[_]]  { self =>
         else seen
     }.reverse
 
+  def distinctBy[A, B: Equal](fa: F[A])(f: A => B): IList[A] =
+    distinctE(fa)(Equal.equalBy(f))
```

Comment:
@TomasMikula the thing is that `seen` is also returned by the method (reversed) so that would require the fold to usea tuple `(List[A], List[B])`, but the gain of not reapplying f again and again could worth it. It probably depends on use-cases. 
On the other hand those methods are O(n*n) anyway so not sure it is worth micro-optimizing the constant factor.